sigmastate/src/main/scala/sigmastate/interpreter/ProverInterpreter.scala

```diff
@@ -171,25 +141,21 @@ trait ProverInterpreter extends Interpreter with AttributionCore {
       val simulated = or.children.forall(_.asInstanceOf[UnprovenTree].simulated)
       or.copy(simulated = simulated)
     case t: CThresholdUnproven =>
-      // If the node is TRESHOLD(k), mark it "real" if at least k of its children are marked real; else mark it "simulated"
-      val c = t.children.foldLeft(0) { (count, child) =>
+      // If the node is THRESHOLD(k), mark it "real" if at least k of its children are marked real; else mark it "simulated"
+      val realCnt = t.children.foldLeft(0) { (count, child) =>
         count + (if (child.asInstanceOf[UnprovenTree].simulated) 0 else 1)
       }
-      t.copy(simulated = c < t.k)
-    case su: UnprovenSchnorr =>
-      // If the node is a leaf, mark it "real'' if the witness for it is available; else mark it "simulated"
-      val secretKnown = secrets.exists {
-        case in: DLogProverInput => in.publicImage == su.proposition
-        case _ => false
-      }
-      su.copy(simulated = !secretKnown)
-    case dhu: UnprovenDiffieHellmanTuple =>
-      // If the node is a leaf, mark it "real" if the witness for it is available; else mark it "simulated"
-      val secretKnown = secrets.exists {
-        case in: DiffieHellmanTupleProverInput => in.publicImage == dhu.proposition
+      t.copy(simulated = realCnt < t.k)
+    // UnprovenSchnorr | UnprovenDiffieHellmanTuple case
+    case ul: UnprovenLeaf =>
+      // If the node is a leaf, mark it "real'' if the witness for it is available
+      // or a hint shows the secret is known to external party participated in multi-signing;
+      // else mark it "simulated"
```

Comment:
Comment better written as 
`If the node is a leaf, mark it "real'' if either the witness for it is available or a hint shows the secret is known to an external participant in multi-signing; else mark it "simulated"`

Reply:
done